### PR TITLE
img_classification/annotation backend

### PR DESCRIPTION
**Pull request by abhishekpratapa**
Image Classification Backend, first implementation completed

Tests:

- [x] test_pass_through
    - Test case where we pass in an annotation SFrame and get back the same SFrame.

- [x] test_get_items
    - Test case where use the api to retrieve items from the SFrame

- [x] test_get_items_out_of_index
    - Test case where we try to retrieve items out of the SFrame however we get an out of index error instead

- [x] test_set_annotations_pass
    - Test case we set the annotations properly with the right format and the annotations are within the index of the SFrame 

- [x] test_set_annotations_out_of_index
    - Test case send annotations out of index insert

- [x] test_set_annotations_wrong_type
    - Test case where we send the wrong type of data

- [x] test_set_annotations_empty
    - Test case send empty annotations

- [x] test_return_annotations
    - Test case we create an SFrame, and return the annotations

- [x] test_return_annotations_drop_na
    - Test case where we return the annotations with the N/A columns dropped

- [x] test_annotation_registry
    - Test case to save the annotation when the user accidentally forgets to assign their annotation to a variable

### CODE REVIEW

File: src/unity/lib/CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,9 @@
 project(unity)
 
-subdirs(visualization)
+subdirs(
+	annotations
```
Comment (znation) [inline src/unity/lib/CMakeLists.txt]: Hmm. Maybe we should consolidate `annotations` and `visualization` into something like `ui`?

Comment (abhishekpratapa) [reply src/unity/lib/CMakeLists.txt]: Good Idea. It'll help considerate the UI code

File: src/unity/lib/annotation/build/format/obj_c/Data.pbobjc.m [hunk 2]
```diff
@@ -0,0 +1,285 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (znation) [inline src/unity/lib/annotation/build/format/obj_c/Data.pbobjc.m]: Why do we have `pbobjc` files generated and committed? Nothing appears to be using them.

File: src/unity/lib/annotation/format/annotate.proto [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    AudioClassificationLabel audioClassificationLabel = 200;
+    DrawingClassificationLabel drawingClassificationLabel = 300;
+    ImageClassificationLabel imageClassificationLabel = 400;
+    ObjectDetectionLabel objectDetectionLabel = 500;
+  }
+}
+
+message Annotation {
+  repeated Label labels = 1;
+  repeated int64 datumHash = 2;
```
[the remark this answers is not in the document]
Comment (abhishekpratapa) [reply src/unity/lib/annotation/format/annotate.proto]: I initially used a hash; however, I realized that it wouldn't work if the user has multiple of the same images in their dataset. So, I actually switched over to using that `datumHash` as a `rowIndex`. I'll replace it to use that wording so it isn't confusing.

File: src/unity/lib/annotation/annotation_base.cpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      break;
+    }
+  });
+}
+
+std::shared_ptr<unity_sframe>
+AnnotationBase::returnAnnotations(bool drop_null) {
+  /* POTENTIAL ISSUE
+   *
+   * Once the user calls returnAnnotations the annotation tool
```
[the remark this answers is not in the document]
Comment (abhishekpratapa) [reply src/unity/lib/annotation/annotation_base.cpp]: Ahh, thanks for the tip! I'll definitely do that!

File: src/unity/lib/CMakeLists.txt [hunk 5]
```diff
@@ -35,18 +38,19 @@ make_library(unity_core
     unity_sframe_builder.cpp
     visualization/columnwise_summary.cpp
   REQUIRES
-    flexible_type
-    pylambda 
+	flexible_type
```
[the remark this answers is not in the document]
Comment (abhishekpratapa) [reply src/unity/lib/CMakeLists.txt]: Fixed Indentation ` [540b6a7](/apple/turicreate/pull/1423/commits/540b6a7989e1321e44da3be3ccf333b9981f9d9e)`

File: src/unity/lib/CMakeLists.txt [hunk 6]
```diff
@@ -1,6 +1,9 @@
 project(unity)
 
-subdirs(visualization)
+subdirs(
+    annotation
+	visualization
```
[the remark this answers is not in the document]
Comment (abhishekpratapa) [reply src/unity/lib/CMakeLists.txt]: Fixed Indentation with ` [540b6a7](/apple/turicreate/pull/1423/commits/540b6a7989e1321e44da3be3ccf333b9981f9d9e)`

File: src/unity/lib/annotation/README.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
+# Annotation
+
+The backend and message formats for TuriCreate Annotation are defined in this directory.
```
[the remark this answers is not in the document]
Comment (abhishekpratapa) [reply src/unity/lib/annotation/README.md]: Fixed in ` [3cb0c44](/apple/turicreate/pull/1423/commits/3cb0c44c1bcc144ebff988681ea92b04e2215dd5)`

File: src/unity/lib/annotation/annotation_base.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (abhishekpratapa) [reply src/unity/lib/annotation/annotation_base.cpp]: Fixed with ` [9ebf1b8](/apple/turicreate/pull/1423/commits/9ebf1b8abfdbbe1b4204224f449f753eef356a03)`

File: src/unity/lib/annotation/format/annotate.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (abhishekpratapa) [reply src/unity/lib/annotation/format/annotate.proto]: Fixed in ` [e51c38b](/apple/turicreate/pull/1423/commits/e51c38b02e8449f87e0bd3139768a56f61aa6418)`

File: src/unity/lib/annotation/annotation_base.cpp [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    : m_data(data), m_data_columns(data_columns),
+      m_annotation_column(annotation_column) {
+  this->_addAnnotationColumn();
+  this->_addIndexColumn();
+  this->_checkDataSet();
+}
+
+void AnnotationBase::annotate(const std::string &path_to_client) {
+  std::shared_ptr<AnnotationBase> self = this->m_self;
+  ::turi::visualization::run_thread([self, path_to_client]() {
```
[the remark this answers is not in the document]
Comment (abhishekpratapa) [reply src/unity/lib/annotation/annotation_base.cpp]: Removed them!

File: src/unity/lib/annotation/build/format/obj_c/Data.pbobjc.m (hunk 2 above)
Comment (abhishekpratapa) [reply src/unity/lib/annotation/build/format/obj_c/Data.pbobjc.m]: Removed the files!

File: src/unity/lib/annotation/annotation_base.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (znation) [reply src/unity/lib/annotation/annotation_base.cpp]: From our in-person discussion, it sounds like we want this to block and return the new SFrame. So we should remove the `run_thread` part here, and make this method non-void (have it return an SFrame). The `while` loop around `aw.good()` will make the calling thread block until the UI is closed (or loses connection).

File: src/unity/lib/annotation/image_classification.cpp [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+ImageClassification::ImageClassification(
+    const std::shared_ptr<unity_sframe> &data,
+    const std::vector<std::string> &data_columns,
+    const std::string &annotation_column)
+    : AnnotationBase(data, data_columns, annotation_column) {
+  /* Since abstract classes cannot use std::make_shared and it needed for the
+   * annotate process, we need the child class to construct the shared pointer
+   * and store it in the protected variables of the inherited abstract
+   * `AnnotationBase` class */
+  m_self = std::make_shared<ImageClassification>(*this);
```
Comment (znation) [inline src/unity/lib/annotation/image_classification.cpp]: The make_shared(*this) pattern is pretty weird. `this` is then also going to be holding a shared pointer to a copy of `this`.

Instead, I would let this class behave normally (just store its own instance members), and make the caller of this constructor use make_shared instead.

File: src/unity/lib/annotation/annotation_base.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (abhishekpratapa) [reply src/unity/lib/annotation/annotation_base.cpp]: Fixed in ` [eb82ffc](/apple/turicreate/pull/1423/commits/eb82ffc4e5ae687f048637a5ec5899a62f57acaa)`

File: src/unity/lib/annotation/image_classification.cpp (hunk 9 above)
Comment (abhishekpratapa) [reply src/unity/lib/annotation/image_classification.cpp]: Fixed in ` [eb82ffc](/apple/turicreate/pull/1423/commits/eb82ffc4e5ae687f048637a5ec5899a62f57acaa)`

**Pull request merged**